server/client.go

```diff
@@ -3636,7 +3640,8 @@ func (c *client) deliverMsg(prodIsMQTT bool, sub *subscription, acc *Account, su
 	// do that accounting here. We only look at client.replies which will be non-nil.
 	if client.replies != nil && len(reply) > 0 {
 		client.replies[string(reply)] = &resp{time.Now(), 0}
-		if len(client.replies) > replyPermLimit {
+		client.repliesSincePrune++
+		if client.repliesSincePrune > replyPermLimit || time.Since(client.lastReplyPrune) > replyPruneTime {
```

Comment:
Do we have a sense under heaby load how much more memory this will hold onto?

Reply:
The original issue was it holding onto too much memory and looping through all of it every message. 

Just added some debug statements and found that if the reply subject is already allowed (eg. `"pub": ">"`), then the reply counter never actually goes up and therefore it never is able to prune that subject out until it expires by time. Just noting this because I'm going to be looking at a fix for that too but not sure if it'll have broader effects (hopefully not).

Assuming that the subjects are getting pruned as they're replied to, at most it should only be able to get the `replies` map to an extra `replyPermLimit` size than what it could've possible been before in the worst case scenario. Even for that to happen it would have to fill the map with subjects, attempt to prune, then expire them all by the next message - in that case, pruning for every message over the `replyPermLimit` would cause it to immediately prune whereas this solution would hold onto that memory for the next `replyPermLimit` messages, making the map size `replyPermLimit * 2`. 

Under normal heavy load it shouldn't make any significant difference, as the current behavior typically should only run the prune once every `replyPermLimit` messages anyway when it's configured properly.